Add postpone and delete buttons to tasks in the task list view.

mbh95 opened the pull request:
Fixes #34 

Postpone opens a popup menu and allows to you add +1h, +1d, or +1w to a task.

Both buttons are hidden until you mouse over a task.

mbh95 committed:
Add postpone and delete actions to tasks in the task list view.

Postpone opens a popup menu and allows to you add +1h, +1d, or +1w to a task.

Both buttons are hidden until you mouse over a task.

anoop901 commented:
The buttons overlap with the text. Can you do something about this?
![Screen Shot 2021-06-16 at 22 53 41](https://user-images.githubusercontent.com/4141909/122328704-b68dfc80-cef5-11eb-86a3-b2fb94fa00e8.png)

anoop901 commented:
Can you put the button on the sidebar too?

anoop901 commented on imports/ui/TaskListItem.tsx:
```diff
@@ -70,6 +112,29 @@ export function TaskListItem({
+          open={Boolean(anchorEl)}
+          onClose={closePostponeMenu}
+        >
+          <MenuItem onClick={postponeBy(1)}>+1h</MenuItem>
```
I'd rather the menu items are not abbreviated, so "+ 1 hour", "+ 1 day", and "+ 1 week".

mbh95 replied:
Done.

anoop901 commented on imports/ui/TaskListItem.tsx:
```diff
@@ -31,13 +50,36 @@ export function TaskListItem({
+  const closePostponeMenu = () => {
+    setAnchorEl(null);
+  };
+  const postponeBy = (hoursToPostponeBy: number) => function () {
```
Please change this to milliseconds, since that interacts more nicely with the date type.

mbh95 replied:
Done.

anoop901 commented on imports/ui/TaskListItem.tsx:
```diff
@@ -31,13 +50,36 @@ export function TaskListItem({
+      return;
+    }
+    const newDate = new Date(task.plannedDate.getTime());
+    newDate.setHours(task.plannedDate.getHours() + hoursToPostponeBy);
```
this can be:
new Date(task.plannedDate.getTime() + millisecondsToPostponeBy)

mbh95 replied:
Done.

anoop901 commented on imports/ui/TaskListItem.tsx:
```diff
@@ -70,6 +112,29 @@ export function TaskListItem({
+          <MenuItem onClick={postponeBy(24)}>+1d</MenuItem>
+          <MenuItem onClick={postponeBy(24 * 7)}>+1w</MenuItem>
+        </Menu>
+        <div className={classes.listItemSecondaryActionSpacer} />
```
Can you put another one of these after the trash icon?

mbh95 replied:
Done.

mbh95 committed:
Addressed PR comments.

- Added text wrapping in the list view so that text can't overlap buttons or go off the side of the screen.
- Refactored PostponeTaskButton into its own component. Left the delete button alone for now.
- Added the postpone button to the side bar.
- Don't show the postpone button (in either place) when planned date isn't set.
- Expanded abbreviated text (e.g. "+1w" -> "+1 week").
- Changed postpone logic to use milliseconds (added constants for ms conversions).
- Added another spacer to the right of the trash icon.

anoop901 commented on imports/ui/TaskConfigInputs.tsx:
```diff
@@ -95,6 +104,7 @@ export function TaskConfigInputs({
             setPlannedDate(newPlannedDate);
           }}
         />
```
The button looks like an oval when it's included here. Also it pushes past the margin. Can you figure out what to do with the styles to fix that?

mbh95 replied:
I still don't really understand why it was becoming an oval, but wrapping it in a div is the easiest fix I found.

anoop901 commented:
the button is oval, and pushes past the margin:

![Screen Shot 2021-06-18 at 11 37 27](https://user-images.githubusercontent.com/4141909/122595001-644ff700-d02d-11eb-9f47-60f52b94ddfa.png)

mbh95 committed:
Changed Postpone button to take in a taskId instead of a postpone function.

mbh95 committed:
Fixed oval button.

mbh95 commented on imports/ui/PostponeTaskButton.tsx:
```diff
@@ -0,0 +1,34 @@
+const MILLIS_PER_DAY = 24 * MILLIS_PER_HOUR;
+const MILLIS_PER_WEEK = 7 * MILLIS_PER_DAY;
+
+export function PostponeTaskButton({ postponeBy }: { postponeBy: (ms: number) => void }): JSX.Element {
```
Done.
